### PR TITLE
[Core] Deploying all-the-protocol-actors

### PR DESCRIPTION
**Pull request by 0xBigBoss**
## Description

Adds necessary updates to begin deploying different kind of actor nodes for Pocket Network. It is still incomplete according to #613 .

## Issue

Fixes https://github.com/pokt-network/pocket/issues/613

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [x] Major breaking change
- [x] Documentation
- [x] Development Tools

## List of changes

- Updates localnet to support deploying different actors
- Adds new config protos for enabling various actors
- Creates and starts new actor specific utility module
- Adds new RPC routes
  - /v1/query/nodeRoles
- Adds new CLI command `bin/p1 query noderoles` for querying a nodes roles
- Stub `HandleRelay`
- Converts "" to ":memory:" to be explicit about using in memory stores
  
## Testing

- [x] `make develop_test`; if any code changes were made
- [x] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [x] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: rpc/handlers_servicer.go
Comment (Olshansk) [inline rpc/handlers_servicer.go:11]: Any reason not to do it in this PR?

File: rpc/handlers_servicer.go
Comment (Olshansk) [inline rpc/handlers_servicer.go:10]:
```suggestion
```

File: rpc/handlers_servicer.go
Comment (Olshansk) [inline rpc/handlers_servicer.go:14]: NIT: I think you can simply this using `slices.IndexFunc` https://stackoverflow.com/a/38654444/768439

File: rpc/v1/openapi.yaml
Comment (Olshansk) [inline rpc/v1/openapi.yaml:94]: Make sure to rebase on top of @h5law's recent changes since I anticipate merge conflicts.

File: utility/portal/module.go
Comment (Olshansk) [inline utility/portal/module.go:1]: Let's remove the portal package for now because we added it to the spec after we started working on V1. 

I anticipate there will be a separate piece of work to add it in (the enum, this package, etc...) in the future, so I want to reduce the confusion.

File: utility/module.go
Comment (Olshansk) [inline utility/module.go:12]: I think we should go with just `fisherman` instead of `fisherman_module`; ditto elsewhere

https://go.dev/blog/package-names

Screenshot 2023-05-03 at 7 09 53 PM

File: utility/module.go
Comment (Olshansk) [inline utility/module.go:119]: Can you move this into a helper function call `validateActorModuleExclusivity` with a `TODO` comment to figure out all the rules we want to enable?

File: utility/module.go
Comment (Olshansk) [inline utility/module.go:122]: We also have a `StringLogArrayMarshaler` helper that might be useful here

File: runtime/configs/proto/fisherman_config.proto
Comment (Olshansk) [inline runtime/configs/proto/fisherman_config.proto:9]: If you can't think of it right now, let's remove the TODOs/INVESTIGATE and add things as we implement e2e features. Ditto for the other configs.

I'm working on those in a notion doc for now

Screenshot 2023-05-03 at 7 36 52 PM

File: utility/validator/module.go
Comment (Olshansk) [inline utility/validator/module.go:33]: As long as you do `_ ValidatorUtility = &validator{}` it'll enforce all the functions

File: utility/validator/module.go
Comment (Olshansk) [inline utility/validator/module.go:13]: I see the approach you took here. Still thinking about, but I do think it's a good start!

File: runtime/configs/config.go
Comment (Olshansk) [inline runtime/configs/config.go:35]: I think we should either remove `Portal` from this PR completely or make sure it's account for everything.

For example:
- `runtime/genesis/proto/genesis.proto`
- `ActorType` in `actor.proto`

Personally, I think one PR to add Portal everywhere later is the way to go

File: utility/fisherman/module.go
Comment (Olshansk) [inline utility/fisherman/module.go:21]: Let's move the `XXInterface` of each actor in `shared/utility_module.go` so its easy to understand and add functionality everywhere.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Looks pretty good to me for the most part. Good foundation to add things to in the future.

Note that you'll have to rebase on top of @h5law's changes (already in) and @okdas has #727 in progress. @gokutheengineer is also going to submit a version of state sync for review soon.

Everything is in 🛫

File: rpc/handlers_servicer.go
Comment (0xBigBoss) [reply rpc/handlers_servicer.go:11]: For sure, I meant to leave it in as a TODO_IN_THIS_COMMIT. I just ran out time before adding the new `GetXXXModule` on the Bus interface to make it easy to do that.

Comment (h5law) [reply rpc/handlers_servicer.go:11]: @Olshansk @0xBigBoss Not sure the purpose of this function but is it not a duplicate of `PostV1ClientRelay` from the RPC spec ima merge ina bit?

File: Makefile
Comment (0xBigBoss) [reply Makefile:326]: ahhh very cool, yes that is actually helpful lol better than googling which command to use on mac vs ubuntu vs alpine

Comment (0xBigBoss) [reply Makefile:326]: > i should learn rust
☕ 🐱

**Review by okdas — COMMENTED** [review]
Please excuse my lack of understanding. While reviewing the config file, I noticed that it seems to allow one process to serve as different actors. Is this the intended behavior? I was under the impression that we were not going to support this functionality.

For example, if one process has fisherman.enabled=true and servicer.enabled=true, would that be considered valid? Or is the configuration file looks that way to follow the module approach, where each module has its own property in the config file?

File: build/localnet/README.md
Comment (0xBigBoss) [reply build/localnet/README.md:162]: I do not believe they will vary greatly. It is more for experimentation or adding specific configurations depending on a developers choice of environment. It's a small escape hatch for when you need to deviate a little bit from the defaults.

I only needed to enable CORS for nodes so I could use swagger. It could be used to change any part of the helm chart.

Perhaps you had something else in mind for customizing one's localnet env?

File: build/localnet/Tiltfile
Comment (0xBigBoss) [reply build/localnet/Tiltfile:209]: They do currently. The actor_number variable increments for each validator, servicer, and fisherman(watcher hehe) which ensures that they do not grab the same key. But they do share all of the keys in the existing `private-keys.yaml` file.

File: build/localnet/manifests/cli-client.yaml
Comment (0xBigBoss) [reply build/localnet/manifests/cli-client.yaml:5]: Isn't the default namespace inherent if you don't specify one with `kubectl` or in your kubeconfig? 

Yes, it's exactly that, I deploy localnet to another namespace for easy cleanup.

File: utility/module.go
Comment (0xBigBoss) [reply utility/module.go:122]: I think they accomplish the same thing.

**Comment by 0xBigBoss** [comment]
> Please excuse my lack of understanding. While reviewing the config file, I noticed that it seems to allow one process to serve as different actors. Is this the intended behavior? I was under the impression that we were not going to support this functionality.
> 
> For example, if one process has fisherman.enabled=true and servicer.enabled=true, would that be considered valid? Or is the configuration file looks that way to follow the module approach, where each module has its own property in the config file?

AFAIK only a validator and servicer can co-exist in the same node. All other roles are exclusive to one another (a validator cannot be a fisherman, a servicer cannot be a fisherman). This will all be enforced in the utility module.

**Comment by Olshansk** [comment]
@0xBigBoss Can you rebase on top of @okdas's PR (#727) once it is merged in?

**Review by Olshansk — COMMENTED** [review]
@0xBigBoss I'm thinking of reviewing and merging the config work (even though it's not prod ready yet) since overlapping work is happening in other PRs (e.g. #789). Wdyt?

**Comment by 0xBigBoss** [comment]
> I'm thinking of reviewing and merging the config work (even though it's not prod ready yet) since overlapping work is happening in other PRs (e.g. #789). Wdyt?

Sounds good @Olshansk. I did my best to merge and rebase the current PR off of main. Also, needed to patch the tests after using the temp config method I added.

File: utility/module.go
Comment (Olshansk) [inline utility/module.go:69]: If you make this a receiver function of the module, there would be no need to bus in any parameters

```go
func (u *utilityModule) enableActorModules() error {
```

File: utility/module.go
Comment (Olshansk) [inline utility/module.go:157]: DItto above.

1. Do not pass in `utilityModule`, make it a receiver
2. Config is already part of the `utilityModule` so access that directly

File: utility/module_enable_actors_test.go
Comment (Olshansk) [inline utility/module_enable_actors_test.go:81]: Use `require` instead of `assert`.

Screenshot 2023-05-31 at 6 43 40 PM

https://github.com/stretchr/testify#require-package

**Review by Olshansk — CHANGES_REQUESTED** [review]
Apologize for the long turnaround on this @0xBigBoss, but PTAL at the comments I left. Everything should be pretty straight forward and the whole team is looking forward to getting this merged in!

File: app/client/cli/query.go
Comment (0xBigBoss) [reply app/client/cli/query.go:904]: I believe I was just following the example in this file. Plus it seems fitting to ouput to stderr on an error.

File: persistence/blockstore/block_store.go
Comment (0xBigBoss) [reply persistence/blockstore/block_store.go:38]: u noticed 😏🤠

**Comment by 0xBigBoss** [comment]
@Olshansk @adshmh I tried to take care of the lil changes, will address and rebase the utlity module changes tom. 🫡🤠

File: app/client/cli/query.go
Comment (h5law) [reply app/client/cli/query.go:904]: @Olshansk yeah this was a change I made when removing the `log` prints to users as stderr for errs makes sense

File: runtime/manager_test.go
Comment (0xBigBoss) [reply runtime/manager_test.go:4245]: Not sure about this one, as the default config is not enabled where as this test is checking a node that is staked as a validator and a servicier.

File: utility/module.go
Comment (0xBigBoss) [reply utility/module.go:157]: We need the root config for this bit, and didn't want to add that as part of the module.

File: build/deployments/docker-compose.yaml
Comment (0xBigBoss) [reply build/deployments/docker-compose.yaml:150]: ah you are right, exit 0 is okay to docker compose here, i confused it with k8s scheduling thinking it would just crash loop.

File: utility/session.go
Comment (Olshansk) [inline utility/session.go:25]: remove?

File: utility/module.go
Comment (Olshansk) [inline utility/module.go:35]: Still TODO?

**Comment by Olshansk** [comment]
@0xBigBoss Left a couple small NITS, but otherwise the changes LGTM as long as the tests pass. I've added the `e2e-devnet-test` label to spin up a devnet for this.

File: utility/module.go
Comment (Olshansk) [inline utility/module.go:146]: OPTIONAL NIT:

I think doing the following is more readable and shorter, but not a blocker

```go
if m, ok := u.actorModules[servicer.ServicerModuleName].(modules.ServicerModule); ok {
	return m, nil
}
u.logger.Warn(...)
...
```

File: utility/module.go
Comment (Olshansk) [inline utility/module.go:154]: Can we ever be in a situation where the map does not contain the key so we get a panic because we try to typeast a nill?

File: utility/servicer/module.go
Comment (Olshansk) [inline utility/servicer/module.go]: Can you move these into `service/relay.go` and make the TECHDEBT more explicit. We are likely going to change all of these but I want to keep deleting them outside the scope of this PR. 

 cc @adshmh for context

File: utility/session.go
Comment (Olshansk) [inline utility/session.go:26]:
```suggestion
```

**Review by Olshansk — CHANGES_REQUESTED** [review]
A few more minor NITS but otherwise LGTM. Will approve after latest changes.

1. @adshmh Please prioritize taking a look at this as it conflicts with your work but I want us to merge it in sooner rather than later.

2. @0xBigBoss Please make sure unit tests and E2E tests are passing (green CI)

File: utility/module.go
Comment (0xBigBoss) [reply utility/module.go:154]: Currently not, but you're right, it is sort of shoot urself in the foot material. Added more checks.

**Comment by 0xBigBoss** [comment]
@Olshansk should be GTG, i had to update some of the stuff for e2e tests due to not using the default namespace.

**Review by Olshansk — APPROVED** [review]
Let's merge it in once CI is all 🟢 !

**Comment by 0xBigBoss** [comment]
Thanks @Olshansk, gonna merge now since the checks failing are for off by one day for the changelogs and the codecov plugin failing on a lot of the code that got merged in from main.

**Pull request merged**
